Refs #31732 -- Fixed django.utils.inspect caching for bound methods.

adamchainz opened the pull request:
Thanks to Simon Charette for the original patch.

From discussion: https://github.com/django/django/pull/13151#issuecomment-801910634

Mogost commented on django/utils/inspect.py:
```diff
@@ -3,14 +3,23 @@
-    return inspect.signature(func)
+def _get_func_parameters(func, remove_self):
+    parameters = tuple(inspect.signature(func).parameters.values())
+    if remove_self:
+        parameters = parameters[1:]
```
@adamchainz 
It says to remove self but can it get called for a static method or a class method?
In the case of classmethod it will remove the cls. In the case of staticmethod it may remove argument.

charettes replied:
Not an issue for `staticmethod` has they are not _methods_ per-se they are mainly just a flag to let Python type building know that the method should not be bound. In other words

```python
class Foo:
    @staticmethod
    def static_bar(arg): pass

inspect.ismethod(Foo.static_bar)  # False
```

You point holds for `classmethod` though but the code works as expected: it's a method so the first argument is ignored if it's bound to an object. We could likely just rename the `remove_self` parameter to `remove_first` to avoid confusion but we certainly add tests to `@classmethod`.

charettes reviewed (commented):
Thanks for submitting a PR for this Adam!

It'd be great to have existence tests for `_get_callable_parameters` to make sure caching works properly.

e.g.
```
self.assertIs(
    _get_callable_parameters(Person().no_arguments),
    _get_callable_parameters(Person().no_arguments),
)
```

I didn't check out the branch locally but from a quick look it seems that none of the added test cover the issue uncovered in #13151.

adamchainz commented:
I renamed `remove_self` to `remove_first` and added a test covering a classmethod, and a test covering the cache returning the same results. Does that test cover the issue you reported @charettes , or is there something else to cover?

charettes reviewed (approved):
That's perfect thanks @adamchainz

Mogost reviewed (approved):
I was originally the one who noticed this issue and I can confirm that after this patch the issue disappears.

felixxm reviewed (approved):
Thanks y'all :rocket: 

I moved extra tests to a separate commit.
